File: dcp-solutions/2024-09-17/solution.php
```php
function countUnivalSubtrees($root) {
    $count = 0;
    function isUnival($node, $val) {
        if (!$node) return true;
        if ($node->val != $val) return false;
        return isUnival($node->left, $val) && isUnival($node->right, $val);
    }

    function dfs($node) use (&$count) {
        if (!$node) return;
        $count += (isUnival($node, $node->val) ? 1 : 0);
        dfs($node->left);
        dfs($node->right);
    }
    dfs($root);
    return $count;
}
```

File: dcp-solutions/2024-09-17/solution.c
```c
int countUnivalSubtrees(struct TreeNode* root) {
    int cnt = 0;
    dfs(root, NULL);
    return cnt;

}

void dfs(struct TreeNode* node, struct TreeNode* val) {
    if (!node)
        return;
    
    if (val && node->val != val->val)
        return;
    
    if (!node->left && !node->right)
        ++cnt;
    
    dfs(node->left, node->val);
    dfs(node->right, node->val);
}
```

This solution works by using a depth-first search (DFS) to traverse the tree. For each node, it checks if all its children have the same value as itself and its parent. If they do, it increments the count of unival subtrees. The DFS also handles the base case where a node is NULL (i.e., we've reached the end of a branch).